presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1102,27 +1105,28 @@ else if (criteria instanceof JoinOn) {
 
             for (Expression conjunct : ExpressionUtils.extractConjuncts((Expression) optimizedExpression)) {
                 conjunct = ExpressionUtils.normalize(conjunct);
-                if (!(conjunct instanceof ComparisonExpression)) {
-                    throw new SemanticException(NOT_SUPPORTED, node, "Non-equi joins not supported: %s", conjunct);
+                if (!isSupportedJoinConjunct(conjunct)) {
+                    throw new SemanticException(NOT_SUPPORTED, node, "Unsupported expression in join clause: %s", conjunct);
                 }
 
-                ComparisonExpression comparison = (ComparisonExpression) conjunct;
```

Comment:
I think the right way to do all of this would be to analyze each conjunct in context of the `output` `RelationType`. Then I think you'll be able to remove all this code with dependency extraction, and split the expressions into left and right